test/jdk/java/awt/regtesthelpers/PassFailJFrame.java

```diff
@@ -258,30 +262,120 @@ private static void getFailureReason() {
-     * window) by the specified position.
+     * window) by the specified position. If testWindow is null, only
+     * the instruction frame is positioned according to the Position
+     * parameter.
      * Note: This method should be invoked from the method that creates
-     * testWindow.
+     * testWindow. At test-level, the testWindow must be made visible
+     * after calling this method.
+     *
+     * @param testWindow test window that the test is created. It can be null.
+     * @param position  position can either be:
+     *                  HORIZONTAL - the test instruction frame is positioned
+     *                  such that its right edge aligns with screen's vertical
+     *                  center and the test window (if not null) is placed to
+     *                  the right of the instruction frame.
+     *
+     *                  VERTICAL - the test instruction frame is positioned such
+     *                  that its bottom edge aligns with the screen's horizontal
+     *                  center and the test window (if not null) is placed below
+     *                  the instruction frame.
      *
-     * @param testWindow test window that the test is created
-     * @param position  position can be either HORIZONTAL (both test
-     *                  instruction frame and test window as arranged
-     *                  side by side) or VERTICAL (both test instruction
-     *                  frame and test window as arranged up and down)
+     *                  TOP_LEFT_CORNER - the test instruction frame is positioned
+     *                  such that its top left corner is at the top left corner of
+     *                  the screen and the test window (if not null) is placed to
+     *                  the right of the instruction frame.
      */
     public static void positionTestWindow(Window testWindow, Position position) {
         Dimension screenSize = Toolkit.getDefaultToolkit().getScreenSize();
+
+        // to get the screen insets inorder to position the frame by taking into
+        // account the location of taskbar/menubars on screen
+        GraphicsConfiguration gc = GraphicsEnvironment.getLocalGraphicsEnvironment()
+                .getDefaultScreenDevice().getDefaultConfiguration();
+        Insets screenInsets = Toolkit.getDefaultToolkit().getScreenInsets(gc);
+
         if (position.equals(Position.HORIZONTAL)) {
             int newX = ((screenSize.width / 2) - frame.getWidth());
-            frame.setLocation(newX, frame.getY());
+            frame.setLocation((newX + screenInsets.left),
+                    (frame.getY() + screenInsets.top));
+            // added to push the updated frame location to window manager
+            // before using it to reposition the test window.
+            Toolkit.getDefaultToolkit().sync();
+            try {
+                Thread.sleep(500);
+            } catch (InterruptedException e) {
+                e.printStackTrace();
+            }
```

Comment:
Is it safe to call on EDT?

If `positionTestWindow` is not to be called on EDT, should it be enforced?

Reply:
The intention is that it can be called either way - on or off.

Reply:
Is that possible? Toolkit thread is not EDT, therefore native events are processed, it's probably enough to update the frame coordinates while EDT is blocked and doesn't handle events.

Reply:
Can you expand on what you mean ? 
What's the problem with native events being processed ? We want that.
And this obviously isn't being called on the toolkit thread.

This change isn't doing anything to that method that might change threading reqts or rules of use.
Any problems already existed. What is being changed (apart from adding the new API) is just to make it more robust.

Reply:
> The intention is that it can be called either way - on or off.

@aivanov-jdk  As @prrace mentioned, the main reason for opting - **Thread.sleep** approach instead of **robot.waitForIdle()** was to give the test user the flexibility to call on or off EDT. (since test cases can contain either AWT or Swing components). With `waitForIdle()` approach we had to wrap parts of the code within EDT and part of it out of EDT, this was confusing and complex.

Reply:
@prrace I'm just trying to understand how it works and whether it changes threading requirements. `Thread.sleep` blocks the EDT, no events are handled. If processing native events is enough to update the frame coordinates, that's good.

Thank you both for clarifying.